scheduler/src/cook/mesos/fenzo.clj

```diff
@@ -0,0 +1,84 @@
+          com.netflix.fenzo.TaskScheduler
+          com.netflix.fenzo.VirtualMachineLease
+          com.netflix.fenzo.plugins.BinPackingFitnessCalculators))
+
+(defn extract-message
+  "For some reason, Fenzo's AssignmentFailure doesn't have a getter for the
+  message field.  Need to parse it from the string representation of the
+  AssignmentFailure."
+  [resource-failure]
+  (last (re-find (re-pattern "message=(\\w+)") (str resource-failure))))
+
+(defn count-resource-failure
+  [totals resource-failure]
+  "Fenzo TaskAssignmentResults tend to provide AssignmentFailures in pairs for each
+   resource placement issue. One AssignmentFailure will describe the resource in the
+   message field; the other will use the Resource field.  We actually want to use
+   the message field because the Enumeration of the Resource field doesn't allow
+   for GPU, whereas the message field can contain the string \"gpu\"."
+  (let [msg (extract-message resource-failure)]
+    (if (seq msg)
+      (update-in totals [:resources msg] (fnil inc 0))
+      totals
+      )))
+
+(defn summarize-placement-failure
+  [totals assignment-result]
+  (let [resources-counted (reduce count-resource-failure totals (.getFailures assignment-result))
+        constraint-failure (.getConstraintFailure assignment-result)]
+    (if constraint-failure
+      (update-in resources-counted [:constraints (.getName constraint-failure)] (fnil inc 0))
+      resources-counted)))
+
+(defn accumulate-placement-failures
+  "Param should be a collection of Fenzo placement failure results for a single
+  job.  Returns a tuple of [job id, single string containing all of the result
+  messages].  Along the way, logs each result message if debug-level logging
+  is enabled."
+  [failures]
+  (let [job (:job  (.. (first failures) (getRequest)))
+        under-investigation? (:job/under-investigation job)
+        summary (when (or under-investigation? (log/enabled? :debug))
+                  (reduce summarize-placement-failure {} failures))]
+    (when (log/enabled? :debug)
+      (log/debug "Job" (:job/uuid job) " placement failure summary:" summary))
+    (when (and under-investigation? (seq summary))
+      [(:db/id job) summary])))
+
+(defn record-placement-failures!
+  "For any failure-results for jobs that are under investigation,  persists all
+  Fenzo placement errors as an attribute of the job.  Also logs information about
+  placement errors if debug-level logging is enabled."
+  [conn failure-results]
+  (log/debug "Fenzo placement failure information:" failure-results)
+  (let [failures-by-job (into {} (map accumulate-placement-failures
+                                      failure-results))
+        transactions (mapv (fn [[job-db-id failures-description]]
+                             {:db/id job-db-id
+                              :job/under-investigation false
+                              :job/last-fenzo-placement-failure (str failures-description)})
```

Comment:
Hi @wyegelwel - I think we have a slight disconnect.  I know that the readers of the API are likely to be non-clojure, and that the API should serve JSON - the good news is, that is already happening!  In the current version of this branch, the summarization data is stored in EDN to the database in a very minimized, "machine-friendly" form that is relatively unlikely to ever need to change.  Then, when a user requests information via the API, this data is transformed to a different structure that is more tailored for human consumption (involving complete sentences etc) and returned as JSON.

Whether we store the data in Datomic as EDN or JSON, it's highly unlikely to ever need truncation, since it will never exceed a few hundred chars (unless we add hundreds of custom constraints that a single job would violate). But if it does happen to get truncated somehow, whether it's EDN or JSON this would likely make it unparseable.